lib/elixir/lib/calendar/datetime.ex

```diff
@@ -930,11 +931,28 @@ defmodule DateTime do
       iex> DateTime.to_iso8601(dt, :basic)
       "20000229T230007-0400"
 
+      iex> dt = %DateTime{year: 2000, month: 2, day: 29, zone_abbr: "AMT",
+      ...>                hour: 23, minute: 0, second: 7, microsecond: {0, 0},
+      ...>                utc_offset: -14400, std_offset: 0, time_zone: "America/Manaus"}
+      iex> DateTime.to_iso8601(dt, :extended, 3600)
+      "2000-03-01T04:00:07+01:00"
+
+      iex> dt = %DateTime{year: 2000, month: 2, day: 29, zone_abbr: "AMT",
+      ...>                hour: 23, minute: 0, second: 7, microsecond: {0, 0},
+      ...>                utc_offset: -14400, std_offset: 0, time_zone: "America/Manaus"}
+      iex> DateTime.to_iso8601(dt, :extended, 0)
+      "2000-03-01T03:00:07+00:00"
+
+      iex> dt = %DateTime{year: 2000, month: 2, day: 29, zone_abbr: "AMT",
+      ...>                hour: 23, minute: 0, second: 7, microsecond: {0, 0},
+      ...>                utc_offset: -14400, std_offset: 0, time_zone: "America/Manaus"}
+      iex> DateTime.to_iso8601(dt, :extended, "Z")
+      "2000-03-01T03:00:07Z"
```

Comment:
This is not quite reversible as `from_iso8601` drops this information, but I can't think of a better way to disambiguate between `0` and `Z`, personally, I would just default `0` to `Z` given that would be reversible for the most common cases, but I believe this will be your preference?

Reply:
My proposal is to not allow “Z” as argument here. Instead, we will only show Z if the offset is zero and the time zone is Etc/UTC. What do you think?

Reply:
If we do this, we can default the offset to zero.

Reply:
Also, to be clear, we should add the given offset into the current offset, instead of discarding the current one. :)

Reply:
I take the current offset into account to get to utc if the datetime is not utc, before converting to the new offset if thats what you mean? Only if the offset is 0 and the timezone is utc do I output `Z` now, did you mean something else? The test below here I also added as it shows my intented use for the library (that is to say, reversibility).